In [6]:
import numpy as np
import pandas as pd
import matplotlib as plt
import statsmodels.api as sm

In [2]:
#(a) read the data
Accord=pd.read_csv('Accord-242A-Spring24.csv')
Accord.info
Accord.head(12)

MonthNumeric MonthFactor  Year  AccordSales  Unemployment  AccordQueries  \
0              1     January  2014        20604           6.6             69   
1              2    February  2014        24622           6.7             74   
2              3       March  2014        33962           6.7             79   
3              4       April  2014        34124           6.2             74   
4              5         May  2014        39637           6.3             75   
5              6        June  2014        32329           6.1             76   
6              7        July  2014        35073           6.2             81   
7              8      August  2014        51075           6.1             80   
8              9   September  2014        32956           5.9             75   
9             10     October  2014        27189           5.7             70   
10            11    November  2014        25275           5.8             69   
11            12    December  2014        31589           5.6             63   

     CPIAll  CPIEnergy  MilesTraveled  
0   235.288    250.340         246531  
1   235.547    249.925         249499  
2   236.028    249.961         251120  
3   236.468    249.864         251959  
4   236.918    249.213         252289  
5   237.231    249.714         252054  
6   237.498    248.744         252111  
7   237.460    245.699         252472  
8   237.477    241.610         253485  
9   237.430    237.061         254117  
10  236.983    229.016         253099  
11  236.252    218.536         255291

In [3]:
#split the data
Accord_train=Accord[(Accord['Year']>=2014) & (Accord['Year']<=2018)]
Accord_test=Accord[(Accord['Year']>=2019) & (Accord['Year']<=2023)]

Accord_train.info()

Accord_train

<class 'pandas.core.frame.DataFrame'>
Int64Index: 60 entries, 0 to 59
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   MonthNumeric   60 non-null     int64  
 1   MonthFactor    60 non-null     object 
 2   Year           60 non-null     int64  
 3   AccordSales    60 non-null     int64  
 4   Unemployment   60 non-null     float64
 5   AccordQueries  60 non-null     int64  
 6   CPIAll         60 non-null     float64
 7   CPIEnergy      60 non-null     float64
 8   MilesTraveled  60 non-null     int64  
dtypes: float64(3), int64(5), object(1)
memory usage: 4.7+ KB


MonthNumeric MonthFactor  Year  AccordSales  Unemployment  AccordQueries  \
0              1     January  2014        20604           6.6             69   
1              2    February  2014        24622           6.7             74   
2              3       March  2014        33962           6.7             79   
3              4       April  2014        34124           6.2             74   
4              5         May  2014        39637           6.3             75   
5              6        June  2014        32329           6.1             76   
6              7        July  2014        35073           6.2             81   
7              8      August  2014        51075           6.1             80   
8              9   September  2014        32956           5.9             75   
9             10     October  2014        27189           5.7             70   
10            11    November  2014        25275           5.8             69   
11            12    December  2014        31589           5.6             63   
12             1     January  2015        21011           5.7             67   
13             2    February  2015        21616           5.5             69   
14             3       March  2015        26018           5.4             76   
15             4       April  2015        27251           5.4             74   
16             5         May  2015        32373           5.6             78   
17             6        June  2015        27477           5.3             79   
18             7        July  2015        34496           5.2             91   
19             8      August  2015        40931           5.1             85   
20             9   September  2015        33641           5.0             75   
21            10     October  2015        30121           5.0             77   
22            11    November  2015        25566           5.1             76   
23            12    December  2015        35056           5.0             72   
24             1     January  2016        20765           4.8             80   
25             2    February  2016        25785           4.9             85   
26             3       March  2016        30523           5.0             84   
27             4       April  2016        31526           5.1             82   
28             5         May  2016        31949           4.8             87   
29             6        June  2016        28806           4.9             87   
30             7        July  2016        31946           4.8             90   
31             8      August  2016        30115           4.9             86   
32             9   September  2016        27204           5.0             81   
33            10     October  2016        25551           4.9             79   
34            11    November  2016        27182           4.7             73   
35            12    December  2016        33783           4.7             75   
36             1     January  2017        19536           4.7             84   
37             2    February  2017        23455           4.6             83   
38             3       March  2017        26824           4.4             87   
39             4       April  2017        26938           4.5             81   
40             5         May  2017        33547           4.4             86   
41             6        June  2017        29791           4.3             94   
42             7        July  2017        30903           4.3            100   
43             8      August  2017        30019           4.4             93   
44             9   September  2017        29789           4.2             80   
45            10     October  2017        26740           4.1             89   
46            11    November  2017        22998           4.2             86   
47            12    December  2017        22115           4.1             90   
48             1     January  2018        17677           4.0             89   
49  

In [4]:
import statsmodels.formula.api as smf
ols = smf.ols(formula='AccordSales ~ Unemployment + AccordQueries + CPIEnergy + CPIAll + MilesTraveled',
                 data=Accord_train)
model1 =ols.fit()
print(model1.summary())


                            OLS Regression Results                            
Dep. Variable:            AccordSales   R-squared:                       0.254
Model:                            OLS   Adj. R-squared:                  0.185
Method:                 Least Squares   F-statistic:                     3.683
Date:                Mon, 12 Feb 2024   Prob (F-statistic):            0.00612
Time:                        16:56:12   Log-Likelihood:                -595.60
No. Observations:                  60   AIC:                             1203.
Df Residuals:                      54   BIC:                             1216.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept      1.735e+05   1.49e+05      1.164

In [7]:
#model is not good, calculate VIF now
from statsmodels.stats.outliers_influence import variance_inflation_factor
def VIF(df, columns):
    values = sm.add_constant(df[columns]).values
    num_columns = len(columns)+1
    vif = [variance_inflation_factor(values, i) for i in range(num_columns)]
    return pd.Series(vif[1:], index=columns)
cols=[ 'Unemployment','AccordQueries', 'CPIEnergy', 'CPIAll','MilesTraveled']
VIF(Accord_train,cols)


Unemployment     31.683094
AccordQueries     1.939028
CPIEnergy        12.340357
CPIAll           45.304922
MilesTraveled    18.510210
dtype: float64

In [8]:
#remove CPIALL because of its high VIF
model2 = smf.ols(formula='AccordSales ~ Unemployment + AccordQueries + CPIEnergy + MilesTraveled',
                 data=Accord_train).fit()
print(model2.summary())
cols=[ 'Unemployment','AccordQueries', 'CPIEnergy','MilesTraveled']
VIF(Accord_train,cols)

                            OLS Regression Results                            
Dep. Variable:            AccordSales   R-squared:                       0.210
Model:                            OLS   Adj. R-squared:                  0.153
Method:                 Least Squares   F-statistic:                     3.660
Date:                Mon, 12 Feb 2024   Prob (F-statistic):             0.0103
Time:                        16:57:22   Log-Likelihood:                -597.33
No. Observations:                  60   AIC:                             1205.
Df Residuals:                      55   BIC:                             1215.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept     -3.448e+04   9.49e+04     -0.363

Unemployment      9.836351
AccordQueries     1.938995
CPIEnergy         1.312659
MilesTraveled    10.051510
dtype: float64

In [9]:
#remove MilesTraveled because of its high VIF
model2 = smf.ols(formula='AccordSales ~ Unemployment + AccordQueries + CPIEnergy ',
                 data=Accord_train).fit()
print(model2.summary())
cols=[ 'Unemployment','AccordQueries', 'CPIEnergy',]
VIF(Accord_train,cols)

                            OLS Regression Results                            
Dep. Variable:            AccordSales   R-squared:                       0.209
Model:                            OLS   Adj. R-squared:                  0.167
Method:                 Least Squares   F-statistic:                     4.938
Date:                Mon, 12 Feb 2024   Prob (F-statistic):            0.00411
Time:                        16:57:23   Log-Likelihood:                -597.36
No. Observations:                  60   AIC:                             1203.
Df Residuals:                      56   BIC:                             1211.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept     -9673.8804   1.44e+04     -0.672

Unemployment     2.245546
AccordQueries    1.891211
CPIEnergy        1.265462
dtype: float64

#### (a)(i)
Accordsales=-9673.8804+4490.1003* Unemployment+230.8945* AccordQueries-13.7786*CPIEnergy
#### (a)(ii)
I run the orignial model first, then get the VIF values for each independent variables and delete the relatively significant large ones. Repeat the process until there no relatively large VPIs.
#### (a)(iii)
Since the R square is pretty low. The model deos not do a good prediction. Therefore, I am not expecting the coefficents make sense. For example, the coefficient for unemployment rate is positive, which means more people loses a job, more cars will be sold. Does not make sense at all. Thus, I am not sure the sighs are correct.
#### (a)(iiii)
The model does not do well to predict training  set observations. R square is only 0.209. Sugguesting the model is weak.

In [10]:
#b
ols = smf.ols(formula='AccordSales ~ MonthFactor+Unemployment + AccordQueries + CPIEnergy + CPIAll + MilesTraveled',
                 data=Accord_train)
model3 =ols.fit()
print(model3.summary())

                            OLS Regression Results                            
Dep. Variable:            AccordSales   R-squared:                       0.748
Model:                            OLS   Adj. R-squared:                  0.654
Method:                 Least Squares   F-statistic:                     7.982
Date:                Mon, 12 Feb 2024   Prob (F-statistic):           2.66e-08
Time:                        16:57:25   Log-Likelihood:                -563.04
No. Observations:                  60   AIC:                             1160.
Df Residuals:                      43   BIC:                             1196.
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept               

#### (b)(i)
AccordSales=9.775* 10^4+762.8155* Unemployment+10.662* AccordQueries-639.8070* CPIALL+0.2497* MilesTraveled-8189.7564* JAN-4825.4928* FEB+430.1538*MAR+5573.2245* MAY+1536.5375* JUN+3894.8622* JUL+8697.6749* AUG+3046.1712* SEP-25.6309* OCT-1493.6532*  NOV+3256.3031* DEC
If Month matcehs, then dummy variable=1, otherwirse 0. For here, there no dummy vairable representing April. Thus=0. For poistive coefficients, people are more likely to buy Accord in such months.
#### (b)(ii)
R square here is 0.748, which is a lot better than the model before. The significant variables are T.August,T.Jan and T.May. They have the samllest p value.
#### (b)(iii)
Definetely, Both the R square and Adjusted R square have increased signifcantly. Thus leads to a much better model.
#### (b)(iv)
sepearte the time into 4 seansons, will make less dummy variables.And I think it will improve the best model since splitting data into seasons makes more intutive sense to carsales.

In [11]:
#since p value for unemployement and AccordQuires is high. remove them will be better.
ols = smf.ols(formula='AccordSales ~ MonthFactor + CPIEnergy + CPIAll + MilesTraveled',
                 data=Accord_train)
model4 =ols.fit()
print(model4.summary())

                            OLS Regression Results                            
Dep. Variable:            AccordSales   R-squared:                       0.748
Model:                            OLS   Adj. R-squared:                  0.669
Method:                 Least Squares   F-statistic:                     9.530
Date:                Mon, 12 Feb 2024   Prob (F-statistic):           2.80e-09
Time:                        16:57:26   Log-Likelihood:                -563.08
No. Observations:                  60   AIC:                             1156.
Df Residuals:                      45   BIC:                             1188.
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept               

In [12]:
def OSR2(model, df_train, df_test, dependent_var):
    y_test=df_test[dependent_var]
    y_pred=model.predict(df_test)
    SSE=np.sum((y_test-y_pred)**2) 
    SST=np.sum((y_test-np.mean(df_train[dependent_var]))**2)
    return 1 -SSE/SST
OSR2(model4, Accord_train, Accord_test, 'AccordSales')

-0.5747827824209855

#### (c)
R square is 0.748 and OSR square is -0.5747827824209855. Since the OSR square value is negative, which means the model does not do a good job on predicitng future sales. Since the test data is based on 2019-2023. There are many events happened may cause a outlier. For example, covid 19 and some wars are happening, which cuases the incrase of price in gas. Thus the model will not be able to do a good prediction on the test data.

#### (d)
I was thinking the change of raw materials will affect accord prices.For example, Steel and Iron. Higher raw material will cause a higher sale price, which leads to lower car sales. Thus I have found the data set from WPU101(iron and steel) online and added them into the dataset.

In [14]:
#read the data
SteelIron=pd.read_csv('WPU101.csv')
SteelIron.head()

DATE  WPU101
0  1926-01-01    11.4
1  1926-02-01    11.4
2  1926-03-01    11.4
3  1926-04-01    11.3
4  1926-05-01    11.3

In [15]:
#change the format
SteelIron['DATE'] = pd.to_datetime(SteelIron['DATE'])
SteelIron['DATE'] = SteelIron['DATE'].dt.strftime('%B %Y')
SteelIron.head()

DATE  WPU101
0   January 1926    11.4
1  February 1926    11.4
2     March 1926    11.4
3     April 1926    11.3
4       May 1926    11.3

In [16]:
#display it the same way as accord data
SteelIron['Year'] = pd.to_datetime(SteelIron['DATE']).dt.year

# Extract only the month name for the 'MonthFactor' column
SteelIron['MonthFactor'] = pd.to_datetime(SteelIron['DATE']).dt.strftime('%B')

# Drop the old 'DATE' column as it's no longer needed
SteelIron = SteelIron.drop(columns=['DATE'])

# Display the first few rows to verify the changes
SteelIron.head()

WPU101  Year MonthFactor
0    11.4  1926     January
1    11.4  1926    February
2    11.4  1926       March
3    11.3  1926       April
4    11.3  1926         May

In [18]:
#select trainning data
SteelIron_train=SteelIron[(SteelIron['Year']>=2014) & (SteelIron['Year']<=2018)]
SteelIron_train

WPU101  Year MonthFactor
1056   238.0  2014     January
1057   234.9  2014    February
1058   231.3  2014       March
1059   234.6  2014       April
1060   233.5  2014         May
1061   232.6  2014        June
1062   232.7  2014        July
1063   233.0  2014      August
1064   233.9  2014   September
1065   230.5  2014     October
1066   226.3  2014    November
1067   224.0  2014    December
1068   222.4  2015     January
1069   209.2  2015    February
1070   205.2  2015       March
1071   200.9  2015       April
1072   198.3  2015         May
1073   200.3  2015        June
1074   198.2  2015        July
1075   193.0  2015      August
1076   190.5  2015   September
1077   181.5  2015     October
1078   175.2  2015    November
1079   172.2  2015    December
1080   173.1  2016     January
1081   172.7  2016    February
1082   175.4  2016       March
1083   184.3  2016       April
1084   192.9  2016         May
1085   193.4  2016        June
1086   193.6  2016        July
1087   195.0  2016      August
1088   191.2  2016   September
1089   186.4  2016     October
1090   189.3  2016    November
1091   196.4  2016    December
1092   204.8  2017     January
1093   207.3  2017    February
1094   213.3  2017       March
1095   212.0  2017       April
1096   213.7  2017         May
1097   213.9  2017        June
1098   215.9  2017        July
1099   214.1  2017      August
1100   215.7  2017   September
1101   211.5  2017     October
1102   208.8  2017    November
1103   212.9  2017    December
1104   217.6  2018     January
1105   222.1  2018    February
1106   227.5  2018       March
1107   234.1  2018       April
1108   239.0  2018         May
1109   242.5  2018        June
1110   244.3  2018        July
1111   244.8  2018      August
1112   244.7  2018   September
1113   244.1  2018     October
1114   246.7  2018    November
1115   245.8  2018    December

In [20]:
#Now we merge the training dataset on month and year
merged_data = pd.merge(SteelIron_train, Accord_train, how='inner', on=['Year', 'MonthFactor'])
merged_data


WPU101  Year MonthFactor  MonthNumeric  AccordSales  Unemployment  \
0    238.0  2014     January             1        20604           6.6   
1    234.9  2014    February             2        24622           6.7   
2    231.3  2014       March             3        33962           6.7   
3    234.6  2014       April             4        34124           6.2   
4    233.5  2014         May             5        39637           6.3   
5    232.6  2014        June             6        32329           6.1   
6    232.7  2014        July             7        35073           6.2   
7    233.0  2014      August             8        51075           6.1   
8    233.9  2014   September             9        32956           5.9   
9    230.5  2014     October            10        27189           5.7   
10   226.3  2014    November            11        25275           5.8   
11   224.0  2014    December            12        31589           5.6   
12   222.4  2015     January             1        21011           5.7   
13   209.2  2015    February             2        21616           5.5   
14   205.2  2015       March             3        26018           5.4   
15   200.9  2015       April             4        27251           5.4   
16   198.3  2015         May             5        32373           5.6   
17   200.3  2015        June             6        27477           5.3   
18   198.2  2015        July             7        34496           5.2   
19   193.0  2015      August             8        40931           5.1   
20   190.5  2015   September             9        33641           5.0   
21   181.5  2015     October            10        30121           5.0   
22   175.2  2015    November            11        25566           5.1   
23   172.2  2015    December            12        35056           5.0   
24   173.1  2016     January             1        20765           4.8   
25   172.7  2016    February             2        25785           4.9   
26   175.4  2016       March             3        30523           5.0   
27   184.3  2016       April             4        31526           5.1   
28   192.9  2016         May             5        31949           4.8   
29   193.4  2016        June             6        28806           4.9   
30   193.6  2016        July             7        31946           4.8   
31   195.0  2016      August             8        30115           4.9   
32   191.2  2016   September             9        27204           5.0   
33   186.4  2016     October            10        25551           4.9   
34   189.3  2016    November            11        27182           4.7   
35   196.4  2016    December            12        33783           4.7   
36   204.8  2017     January             1        19536           4.7   
37   207.3  2017    February             2        23455           4.6   
38   213.3  2017       March             3        26824           4.4   
39   212.0  2017       April             4        26938           4.5   
40   213.7  2017         May             5        33547           4.4   
41   213.9  2017        June             6        29791           4.3   
42   215.9  2017        July             7        30903           4.3   
43   214.1  2017      August             8        30019           4.4   
44   215.7  2017   September             9        29789           4.2   
45   211.5  2017     October            10        26740           4.1   
46   208.8  2017    November            11        22998           4.2   
47   212.9  2017    December            12        22115           4.1   
48   217.6  2018     January             1        17677           4.0   
49   222.1  2018    February             2        19753           4.1   
50   227.5  2018       March             3        24171           4.0   
51   234.1  2018       April             4        21751           4.0   
52   239.0  2018         May             5        28212           3.8   
53   242.5  2018        June             6        26726       

In [21]:
#Now we do the new model based on (c)
ols = smf.ols(formula='AccordSales ~ MonthFactor + CPIEnergy + CPIAll + MilesTraveled+WPU101',
                 data=merged_data)
model5 =ols.fit()
print(model5.summary())

                            OLS Regression Results                            
Dep. Variable:            AccordSales   R-squared:                       0.758
Model:                            OLS   Adj. R-squared:                  0.675
Method:                 Least Squares   F-statistic:                     9.172
Date:                Mon, 12 Feb 2024   Prob (F-statistic):           4.02e-09
Time:                        17:16:36   Log-Likelihood:                -561.88
No. Observations:                  60   AIC:                             1156.
Df Residuals:                      44   BIC:                             1189.
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept               

#### (d)
The new model increased the R square value by 0.1. Not too much, but it improved the model a little. Overall, it is a great news.And the new factor, WPU101 has a low p value, which means it is significant to the accordsales. Thus it added a useful feature for predicting future sales and it makes sense. The coefficent for WPU101 is nrgative. Suggueting that as the price of Steel and Iron goes up, people are less willing to buy a car because of the higher price of Accord.
